cluster_regional_private.tf

```diff
@@ -0,0 +1,153 @@
+/**
+ * Copyright 2018 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/******************************************
+  Create regional cluster
+ *****************************************/
+resource "google_container_cluster" "primary_private" {
```

Comment:
What is the reason for needing a separate cluster config for private clusters? Ideally we should minimize how many different `google_container_cluster` resources we have?

Reply:
Apologies, a commit was missing from this. https://github.com/terraform-google-modules/terraform-google-kubernetes-engine/pull/51/commits/719ee8ba49e9cbc62b5c8454aff8cc01719b9780 shows the actual configuration of the private cluster.

The issue here is that Terraform 0.11.x doesn't support conditionally adding this block. Hopefully Terraform 0.12 will make this less painful, but for now I'm not sure there's a better option than duplicating the resource.

Reply:
As far as I can tell, the `private_cluster_config` block itself doesn't force anything. If you set all its arguments to false, you would have a standard public cluster.

it's the `enable_private_nodes` piece which is what is primarily meant by a private cluster usually.

Reply:
@morgante Unfortunately, the provider requires a `master_ipv4_cidr_block` to be [provided as a valid CIDR](https://github.com/terraform-providers/terraform-provider-google/blob/91a513ab832f35673720f5f5902e1f6f34841169/google/resource_container_cluster.go#L567) whenever the `private_cluster_config` block is specified, and overriding it from the default (unspecified) value can have hard-to-reason-about results. If we simply don't provide that attribute, things work, but we fall into the Terraform 0.11.x trap where an empty string is not considered the same as null, and we have no way to represent null when it should be so.

I would suggest that we merge this as-is, understanding that there is additional complexity as a result of this, and file an upstream issue to make the validation on `master_ipv4_cidr_block` less stringent.

Reply:
I think we should hold off as this is a change which will cause breakage later on if we walk it back and instead file an upstream provider issue.

Reply:
Rather than driving optional features through variable values, can we provide a submodule which can be optionally included in a configuration if a private cluster is desired?

Reply:
Let's approach this by making a submodule as @aaron-lane suggested.